File: examples/15_algorithms/15f_input/inversions.c
/* The same problem as inversions.ppy, reading the judge's input with scanf. */
#include <stdio.h>
#include <stdlib.h>

#define LIMIT 1000004

static void add(long long *tree, long long position, long long delta) {
    long long index = position + 1;
    while (index < LIMIT) {
        tree[index] += delta;
        index += index & -index;
    }
}

static long long prefix(const long long *tree, long long position) {
    long long total = 0;
    long long index = position + 1;
    while (index > 0) {
        total += tree[index];
        index -= index & -index;
    }
    return total;
}

static long long count_inversions(const long long *values, long long size, long long *tree) {
    long long total = 0;
    for (long long i = size - 1; i >= 0; i--) {
        total += prefix(tree, values[i] - 1);
        add(tree, values[i], 1);
    }
    return total;
}

int main(void) {

    long long size = 0;
    long long *values = NULL;
    if (scanf("%lld", &size) == 1) {
        values = malloc((size_t)size * sizeof(long long));
        for (long long i = 0; i < size; i++) {
            if (scanf("%lld", &values[i]) != 1) {
                return 1;
            }
        }
    } else {
        size = 500000;
        values = malloc((size_t)size * sizeof(long long));
        for (long long i = 0; i < size; i++) {
            values[i] = (i * 7919) % 1000003;
        }
    }

    long long *tree = calloc(LIMIT, sizeof(long long));
    long long answer = count_inversions(values, size, tree);

    printf("%lld\n", answer);
    return 0;
}

File: examples/15_algorithms/15f_input/inversions.py
"""Counting inversions with a Fenwick tree.

Input:  N on the first line, then N integers.
Output: how many pairs are out of order.
"""

import array

LIMIT = 1000004


def add(tree, position, delta):
    index = position + 1
    while index < LIMIT:
        tree[index] += delta
        index += index & -index
    return index


def prefix(tree, position):
    total = 0
    index = position + 1
    while index > 0:
        total += tree[index]
        index -= index & -index
    return total


def count_inversions(values, tree):
    total = 0
    for i in range(len(values) - 1, -1, -1):
        total += prefix(tree, values[i] - 1)
        add(tree, values[i], 1)
    return total


def read_values():
    """The judge's input, or a generated stand-in when nothing is piped in."""
    try:
        count = int(input())
    except EOFError:
        return array.array("q", [(i * 7919) % 1000003 for i in range(500000)])
    values = array.array("q", [0] * count)
    for i in range(count):
        values[i] = int(input())
    return values


def main():
    values = read_values()
    tree = array.array("q", [0] * LIMIT)
    print(count_inversions(values, tree))


main()
